tests/test_engine.py

```diff
@@ -364,3 +364,46 @@ def test_invalid_file_charset():
         )
 
     assert "Unknown encoding: 'not-a-real-encoding'" == str(exc_info.value)
+
+
+def test_render_to_string_success():
+    rusty_engine = engines["rusty"]
+
+    # Test with single template name
+    result = rusty_engine.engine.render_to_string("basic.txt", {"user": "Alice"})
+    assert result == "Hello Alice!\n"
+
+    # Test with template list/tuple
+    result = rusty_engine.engine.render_to_string(
+        ["nonexistent.html", "basic.txt"], {"user": "Bob"}
+    )
+    assert result == "Hello Bob!\n"
+    result = rusty_engine.engine.render_to_string(
+        ("nonexistent.html", "basic.txt"), {"user": "Bob"}
+    )
+    assert result == "Hello Bob!\n"
+
+    # Test with no context
+    result = rusty_engine.engine.render_to_string("basic.txt")
+    assert result == "Hello !\n"
+
+
+@pytest.mark.parametrize(
+    "template_name", [{"nonexistent.html", "basic.txt"}, [1, 2], None]
+)
+def test_render_to_string_invalid_template_name(template_name):
+    rusty_engine = engines["rusty"]
+
+    with pytest.raises(TypeError, match="object cannot be cast as 'str'"):
+        rusty_engine.engine.render_to_string(template_name, {"user": "Bob"})
```

Comment:
I think I'd prefer to be more explicit about the error messages:

```suggestion
@pytest.mark.parametrize(
    "template_name,message",
    [
        ({"nonexistent.html", "basic.txt"}, "'set' object cannot be cast as 'str'"),
        ([1, 2], "'list' object cannot be cast as 'str'"),
        (None, "'NoneType' object cannot be cast as 'str'"),
    ],
)
def test_render_to_string_invalid_template_name(template_name, message):
    rusty_engine = engines["rusty"]
    with pytest.raises(TypeError) as exc_info:
        rusty_engine.engine.render_to_string(template_name, {"user": "Bob"})

    assert str(exc_info.value) == message
```

I often struggle to remember that `match` is treated as a regex.